bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+                        msg: ConnToMainMsgInner::FailedServerHandshake { addr },
+                    })
+                    .await;
+                return;
+            }
+        };
+
+        // Inform the main task that we have connected to a peer
+        let _ = main_tx
+            .send(ConnToMainMsg {
+                handle_unique_id: unique_id,
+                msg: ConnToMainMsgInner::ConnectedServer {
+                    accepted_addr: addr,
+                    addr: identify.addr.clone(),
+                    peer_id: identify.id.clone(),
+                },
+            })
+            .await;
+
+        let mut conn = EstablishedConn::new(
+            identify.id.clone(),
+            unique_id,
+            write_sock,
+            read_sock,
+            main_tx.clone(),
+            rx,
+            log.clone(),
+        );
+
+        // We always exit on server tasks, as the remote peer
+        // will reconnect.
+        let _ = conn.run().await;
+    });
+
+    AcceptedConnHandle { handle, tx, addr, unique_id }
+}
+
+// Serialize and write `msg` into `buf`, prefixed by a 4-byte big-endian size header
+//
+// Return the total amount of data written into `buf` including the 4-byte header
+fn write_framed<T: Serialize + ?Sized>(
+    msg: &T,
+) -> Result<Vec<u8>, ciborium::ser::Error<std::io::Error>> {
+    let mut cursor = Cursor::new(vec![]);
+    // Write a size placeholder
+    std::io::Write::write(&mut cursor, &[0u8; FRAME_HEADER_SIZE]);
+    cursor.set_position(FRAME_HEADER_SIZE as u64);
+    ciborium::into_writer(msg, &mut cursor)?;
+    let size: u32 =
+        (cursor.position() - FRAME_HEADER_SIZE as u64).try_into().unwrap();
+    let mut buf = cursor.into_inner();
+    buf[0..FRAME_HEADER_SIZE].copy_from_slice(&size.to_be_bytes());
+    Ok(buf)
+}
+
+// Decode the 4-byte big-endian frame size header
+fn read_frame_size(buf: [u8; FRAME_HEADER_SIZE]) -> usize {
+    u32::from_be_bytes(buf) as usize
+}
+
+#[derive(Debug, From)]
+enum HandshakeError {
+    Serialization(ciborium::ser::Error<std::io::Error>),
+    Deserialization(ciborium::de::Error<std::io::Error>),
+    Io(tokio::io::Error),
+    UnsupportedScheme,
+    UnsupportedVersion,
+    Timeout,
+}
+
+// Perform scheme/version negotiation and exchange peer_ids for scheme v0
+async fn perform_handshake(
+    sock: TcpStream,
+    local_peer_id: &Baseboard,
+    local_addr: SocketAddrV6,
+    log: &Logger,
+) -> Result<(OwnedReadHalf, OwnedWriteHalf, Identify), HandshakeError> {
+    // Enough to hold the `Hello` and `Identify` messages
+    let mut read_buf = [0u8; 128];
+    let (mut read_sock, mut write_sock) = sock.into_split();
+
+    // Serialize and write the handshake messages into `write_buf`
+    let mut hello_cursor = Cursor::new(Hello::default().serialize());
+    let identify = write_framed(&Identify {
+        id: local_peer_id.clone(),
+        addr: local_addr,
+    })?;
+    let mut identify_cursor = Cursor::new(&identify);
```

Comment:
Having a separate buffer / cursor for `Hello` and `identity` seems awkward and requires extra tracking; could you serialize them back-to-back in a single buffer instead?

```rust
    // Serialize and write the handshake + identity messages to a local buffer
    let out: Vec<u8> = Hello::default()
        .serialize()
        .into_iter()
        .chain(write_framed(&Identify {
            id: local_peer_id.clone(),
            addr: local_addr,
        })?)
        .collect();
    let mut out_cursor = Cursor::new(&out);
```

this requires a few changes later on:
```diff
@@ -585,9 +588,7 @@ async fn perform_handshake(

         let end = initial_read + identify_len;

-        let hello_written = !hello_cursor.has_remaining();
-
-        if identify.is_some() && !identify_cursor.has_remaining() {
+        if identify.is_some() && !out_cursor.has_remaining() {
             return Ok((read_sock, write_sock, identify.unwrap()));
         }

@@ -595,12 +596,8 @@ async fn perform_handshake(
             _ = sleep(timeout) => {
                 return Err(HandshakeError::Timeout);
             }
-            _ = write_sock.writable(), if identify_cursor.has_remaining() => {
-                if hello_cursor.has_remaining() {
-                    write_sock.write_buf(&mut hello_cursor).await?;
-                } else {
-                    write_sock.write_buf(&mut identify_cursor).await?;
-                }
+            _ = write_sock.writable(), if out_cursor.has_remaining() => {
+                write_sock.write_buf(&mut out_cursor).await?;
             }
             res = read_sock.read(&mut read_buf[total_read..end]), if identify.is_none() => {
                 let n = res?;
```

Reply:
I totally agree. I definitely had them in a single buffer at one point, but via iteration ended up with this weird mess. I'll do as you suggest.

Reply:
Fixed in 7a9167133